CCV spec draft w/ Validator Set Update (#48)
informalsystems/cross-chain-validation: ics-028-cross-chain-validation/README.md

- The *completion*, e.g., the tokens are actually unbonded and transferred back to the delegator.  To complete, unbonding operations must reach *maturity*, i.e., `UnbondingPeriod` must elapse since the operations were initiated.  

### Definitions

The following Figure shows an overview of CCV together with the new terms and concepts introduced.

![CCV Overview](./ccv-vsc-overview.png?raw=true)

**Provider Chain**: The blockchain that provides security, i.e., manages the validator set of the consumer chain.

**Consumer Chain**: The blockchain that consumes security, i.e., enables the provider chain to correctly manage its validator set.

> Note that in the current version the validator set of the consumer chain is entirely provided by the provider chain. 

**Validator Set Change (VSC)**: A change in the validator set of the provider chain that must be reflected in the validator set of the consumer chain. It consists of a batch of validator updates, i.e., changes in the voting powers granted to validators on the provider chain and, due to CCV, also on the consumer chain.

> **Matured VSC**: A VSC that has matured on the consumer chain, i.e., all the unbonding operations that resulted in validator updates included in the VSC have matured on the consumer chain. 

Reviewer: I would say, "...., i.e., for each validator that was removed from the validator set due VSC, the unbonding period on the consumer chain elapsed.

Author: This is not completely true: An unbonding operation does not necessarily entail that the validator is removed. For example, a delegator unbonding some of its stake it's an unbonding operation, but if the unbonded amount it's not too large, the validator will most likely remain in the set.

First, we define the properties for the CCV channel. Then, we define the guarantees provided by CCV.

- **Channel Uniqueness**: The provider and consumer chains communicate over a unique channel.
- **Channel Validity**: If a packet P is received by one end of the channel, then P MUST have been sent by the other end of the channel.
- **Channel Order**: If a packet P1 is sent over the channel before a packet P2, then P2 MUST NOT be received by the other end of the channel before P1. 
- **Channel Liveness**: Every packet sent over the channel MUST eventually be received by the other end of the channel. 

> Note that the following properties are concerned with only one provider chain and one consumer chain. 

> TODO: Do we need properties for multiple consumer chains? What the interaction between two consumer chains connected to the same provider chain?

CCV is based on the following system invariants:
- The voting power granted to validators on the consumer chain corresponds to the amount of tokens bonded by these validators on the provider chain.
- The unbonding of any amount of the tokens bonded on the provider chain MUST NOT happen before it matures on the consumer chain.

> 

Reviewer: to do: add before matures on all consumer chains.

### Definitions and Overview

This section defines the new terms and concepts introduced by CCV and provides an overview of CCV.

**Provider Chain**: The blockchain that provides security, i.e., manages the validator set of the consumer chain.

**Consumer Chain**: The blockchain that consumes security, i.e., enables the provider chain to correctly manage its validator set.

> Note that in the current version the validator set of the consumer chain is entirely provided by the provider chain.

**CCV Module**: The module that implements the CCV protocol.

**CCV Channel**: A unique, ordered IBC channel (as define in [ICS 4](../../core/ics-004-channels-and-packets-semantics)) that is used by the two CCV modules to exchange messages.

> Note that the IBC handler interface & IBC relayer module interface are as defined in [ICS 25](../../core/ics-025-handler-interface) and [ICS 26](../../core/ics-026-routing-module), respectively.

CCV must handle the following types of operations:
- **Channel Initialization**: Create a unique, order IBC channel between the provider chain and the consumer chain.

> - **Validator Set Update**: Update the validator set of the consumer chain based on the tokens bonded by validators on the provider chain and manage *unbonding operations*, i.e., operations of unbonding bonded tokens.

Author: > What do you want to say with "based on the tokens bonded by validators ". It reads like the computation of who is in or out of the validator set is done in CCV and not in Staking.
What about the following: "Update the validator set of the consumer chain based on information received from the Staking module on the amount of tokens bonded by validators on the provider chain" ?

> "manage *unbonding operations", i.e., operations of unbonding bonded tokens" -> CCV does not unbound any tokens. It only measures if the unbonding period elapsed (on the consumer chain). The staking does the token unbonding, after checking with CCV that the unbonding period elapsed.
That's why I said "manage" instead of unbond. The CCV manages the unbonding by delaying the staking module. What about the following: "... and enable the timely completion of *unbonding operations*, i.e., operations of unbonding bonded tokens."

#### Channel Initialization Overview

The following Figure shows an overview of the CCV Channel initialization. 

![Channel Initialization Overview](./ccv-init-overview.png?raw=true)

Consumer chains are created through governance proposals. For details on how governance proposals work, take a look at the [Cosmos SDK documentation](https://docs.cosmos.network/master/modules/gov/).

The channel initialization consists of four phases:

> - **Create clients**: The CCV module on the provider chain handles every passed proposal to spawn a new consumer chain. Once it receives a proposal, the CCV module creates a client of the consumer chain. Then, every validator in the validator set of the provider chain must start a full node of the consumer chain. Once the consumer chain starts, the `InitGenesis()` method of the CCV module is invoked (for more details, take a look at the [Cosmos SDK documentation](https://docs.cosmos.network/master/building-modules/genesis.html)) and a client of the provider chain is created. Note that at  genesis, the validator set of the consumer chain matches the validator set of the provider chain.

Reviewer: Add that the client initialisation (on the consumer) uses the provider's state and consensus state from genesis.  It's a CCV thing different from others, e.g. TT. 

Please make clear what genesis should contain.

Note that the "*FIRST*" keyword in the above description ensures the uniqueness of the IBC channel.

#### Validator Set Update Overview

The following Figure shows an overview of the Validator Set Update operation of CCV. 

![Validator Set Update Overview](./ccv-vsc-overview.png?raw=true)

**Validator Set Change (VSC)**: A change in the validator set of the provider chain that must be reflected in the validator set of the consumer chain. It consists of a batch of validator updates, i.e., changes in the voting powers granted to validators on the provider chain and, due to CCV, also on the consumer chain.

> **Background**: In the context of single-chain validation, the changes of the validator set are triggered by the Staking module. For more details, take a look at the [Cosmos SDK documentation](https://docs.cosmos.network/master/modules/staking/). 

> **Matured VSC**: A VSC that has matured on the consumer chain, i.e., a certain period of time, known as the *unbonding period* (i.e., `UnbondingPeriod`) has elapsed since the VSC was received by the consumer chain. The intuition behind a VSC reaching maturity on a consumer chain is that all the unbonding operations that resulted in validator updates included in that VSC have matured on the consumer chain. 

Author: It is not circular: "all the unbonding operations that resulted in validator updates included in that VSC have matured" not the same as "VSC that have matured".

- first, given any two chains A and B, we cannot assume that A's rate of adding new block is the same as B's rate (i.e., we consider the sequences of blocks of any two chains to be completely asynchronous); 
- and second, due to relaying delays, we cannot assume that the rate of sending VSCs matches the rate of receiving VSCs.

As a result, is it possible for multiple VSCs to be received by the consumer chain within the same block and be applied together at the end of the block, i.e., the validator updates within the VSCs are being *aggregated* by keeping only the latest update per validator. As a consequence, some validator sets on the provider chain are not existing on the consumer chain. In other words, the validator sets on the consumer chain form a *subsequence* of the validator sets on the provider chain. Nonetheless, as a **requirement of CCV**, *all the validator updates on the provider chain MUST be included in the sequence of validator sets on the consumer chain*. 

This is possible due to the *update inclusion* property, which is a consequence of every validator update containing the absolute voting power of that validator.

> - **Update Inclusion**: Let U1 and U2 be two updates of the voting power of a validator V. If U1 occurs before U2, then U2 contains U1. 

Reviewer: Can you please define the inclusion relation ?

Author: I'm not sure I understand what you mean by "inclusion relation". I tried to clarify the property. Let me know what you think.

As a result, it is sufficient for only the last update per validator to be applied. Since the last update of a validator contains all the previous updates of that validator, once it is applied, all the previous updates are also applied.

---

#### Staking Module Interface

> The following properties define the guarantees of *requesting* VSCs to the consumer chains as a consequence of validator updates on the provider chain. 

Reviewer: Whose guarantees ? Staking's of CCV ? Some look like assumptions that CCV makes on its calling context.

- **Validator Update to VSC Validity**: Every VSC requested to a consumer chain MUST contain only validator updates that resulted from a change in the amount of bonded tokens on the provider chain.

> - **Validator Update to VSC Order**: Let U1 and U2 be two validator updates on the provider chain. If U1 occurs before U2, the U2 MUST NOT be included in a requested VSC before U1. 

Reviewer: the -> then

The following properties define the guarantees of *applying* on the consumer chain VSCs *requested* by the provider chain.  
- **Apply VSC Validity**: Every VSC applied by the consumer chain MUST be requested by the provider chain.
- **Apply VSC Order**: If a VSC C1 is requested by the provider chain before a VSC C2, then the consumer chain MUST NOT apply the validator updates included in C2 before the validator updates included in C1.
- **Apply VSC Liveness**: If the provider chain requests a VSC C, then the consumer chain MUST eventually apply all validator updates included in C.

The following properties define the guarantees of *registering* on the provider chain maturity notifications (from the consumer chain) of VSCs *requested* by the provider chain to the consumer chain.
- **Register Maturity Validity**: If the provider chain registers a maturity notification of a VSC from the consumer chain, then the provider chain MUST have requested that VSC to the consumer chain. 

> - **Register Maturity Timeliness**: If the provider chain requests a VSC C to the consumer chain, then the provider chain MUST NOT register a maturity notification of C before `UnbondingPeriod` has elapsed on the consumer chain since the consumer chain received C.

Reviewer: since the consumer chain received and applied another VSC after C ?

Author: This is actually not true at the moment in the implementation. The timer for sending the ACK for a VSCPacket is started when the packet is received. For the second invariant to hold, it should indeed start when the VSC is applied. Right?

Should we change this in the spec and then ask the dev side to change it also in the code?

```typescript
// implements the ICS26 interface
function onChanOpenInit(
  order: ChannelOrder,
  connectionHops: [Identifier],
  portIdentifier: Identifier,
  channelIdentifier: Identifier,
  counterpartyPortIdentifier: Identifier,
  counterpartyChannelIdentifier: Identifier,
  version: string) {
    // the channel handshake MUST be initiated by consumer chain
    abortTransactionUnless(FALSE)
}
```
- **Initiator:** 
  - The IBC module on the provider chain.
- **Expected precondition:**
  - The IBC module on the provider chain received a `ChanOpenInit` message on a port the provider chain CCV module is bounded to.
- **Expected postcondition:** 

>   - None.

Author: I think it's understood from None. We should aim to make the spec as less verbose as possible. It's already too verbose since it's an english spec.

- **Channel Validity:** Follows directly from the Valid Blockchain assumption.

- **Channel Order:** The provider chain accepts only ordered channels when receiving a `ChanOpenTry` message (cf. Valid Blockchain). Similarly, the consumer chain accepts only ordered channels when receiving `ChanOpenInit` messages (cf. Valid Blockchain). Thus, the property follows directly from the fact that the CCV channel is ordered. 

- **Channel Liveness:** The property follows from the Correct Relayer assumption. 

- **Apply VSC Validity:** The property follows from the following two assertions.
  - The consumer chain only applies VSCs received in `VSCPacket`s through the CCV channel (cf. Valid Blockchain).
  - The provider chain only sends `VSCPacket`s containing requested VSCs (cf. Valid Blockchain). 

> - **Apply VSC Order:** We prove the property through contradiction. We assume the consumer chain applies the validator updates included in C2 before the validator updates included in C1. The following sequence of assertions leads to a contradiction.

Reviewer: Given two validator sets C1 and C2 such that on the provider chain C2 follows C1, "we assume that the consumer chain applies ... etc "

- **Apply VSC Liveness:** The provider chain eventually sends over the CCV channel a `VSCPacket` containing C (cf. Valid Blockchain). As a result, the consumer chain eventually receives this packet (cf. Channel Liveness). Then, the consumer chain aggregates all received VSCs at the end of the block and applies all the aggregated updates (cf. Valid Blockchain). As a result, the consumer chain applies all validator updates in C (cf. Update Inclusion).

- **Register Maturity Validity**: The property follows from the following sequence of assertions.
  - The provider chain only registers VSC maturity notifications when receiving on the CCV channel acknowledgements of `VSCPacket`s (cf. Valid Blockchain). 
  - The provider chain receives on the CCV channel only packets sent by the consumer chain (cf. Channel Validity).
  - The consumer chain only acknowledges `VSCPacket`s that it receives on the CCV channel (cf. Valid Blockchain).
  - The consumer chain receives on the CCV channel only packets sent by the provider chain (cf. Channel Validity). 
  - The provider chain only sends `VSCPacket`s containing requested VSCs (cf. Valid Blockchain). 

> - **Register Maturity Timeliness:** We prove the property through contradiction. We assume that the provider chain registers a maturity notification of C before `UnbondingPeriod` has elapsed on the consumer chain since the consumer chain received C. The following sequence of assertions leads to a contradiction.

Reviewer: why "since the consumer chain received C" ? you do not start the unbonding of a val set V when you install V but when you receive a new val set.

Author: That's a bit unclear at the moment. Currently, in the implementation the unbonding period starts from the time the VSC was **received**.  The properties are consistent with this approach. 

However, I think the unbonding time should start from the time the VSC was **applied**. In practice there is not much difference since between receiving and applying a VSC are just a few seconds (the time of creating a new block). In theory though, that's incorrect. **The unbonding period should start from the time validators are part of the validator set of the consumer chain** (i.e., the VSC is applied).

---
ics: 28
title: Cross-Chain Validation
stage: draft
category: IBC/APP
requires: 25, 26
kind: 
author: 
created: 
modified: 
---

## Synopsis

This standard document specifies packet data structure, state machine handling logic, and encoding details for Cross-Chain Validation (CCV). Note that CCV is the specific IBC level protocol that enables *Interchain Security*, a Cosmos-specific category of *Shared Security*.

At a high level, CCV enables a *provider chain* (e.g., the Cosmos Hub) to provide *security* to a *consumer chain*. Before we go further, we give a short overview of what we mean by security in the context of Cosmos. 

> **Security Model**: We consider chains that reach consensus through a proof of stake mechanism based on the model of [weak subjectivity](https://blog.ethereum.org/2014/11/25/proof-stake-learned-love-weak-subjectivity/). The next block in a blockchain is *validated* and *voted* upon by a set of pre-determined *full nodes*, also known as *validators*. We refer to the validators eligible to validate a block as that block's *validator set*. To be part of the validator set, a validator needs to *bond* (i.e., lock, stake) an amount of tokens for a minimum period of time, known as the *unbonding period*. The amount of tokens bonded gives a validator's *voting power*. If a validator misbehave (e.g., validates two different blocks at the same height), its bonded tokens can be slashed. Note that the unbonding period enables the system to punish a misbehaving validator after the misbehavior is committed. For more details, take a look at the [Tendermint Specification](https://github.com/tendermint/spec/blob/master/spec/core/data_structures.md) and the [Light Client Specification](https://github.com/tendermint/spec/blob/master/spec/light-client/verification/verification_002_draft.md#part-i---tendermint-blockchain).

Reviewer 2: Typo: "If a validator misbehaves (e.g., validates two different blocks ...")

In the context of CCV, the validator set of the consumer chain is chosen based on the tokens validators bonded on the provider chain, i.e., is chosen from the validator set of the provider chain. When these validators misbehave on the consumer chain, their bonded tokens on the provider chain is slashed. As a result, the security gained from the value of the bonded tokens on the provider chain is shared with the consumer chain. For more details, take a look at the [Interchain Security light paper](https://github.com/cosmos/gaia/blob/main/docs/interchain-security.md).

> The exchange of messages between the provider and consumer chains is done through the IBC protocol over an *unique*, *ordered* channel. 

Reviewer 2: Typo: "... is done through the IBC protocol over a *unique*, ...".

> Throughout this document, we will use the terms chain and blockchain interchangeably.

### Motivation

CCV is a primitive (i.e., a building block) that enables arbitrary shared security models: The security of a chain can be composed of security transferred from multiple provider chains including the chain itself (a consumer chain can be its own provider). As a result, CCV enables chains to borrow security from more established chains (e.g., Cosmos Hub), in order to boost their own security, i.e., increase the cost of attacking their networks. 
> **Intuition**: For example, for chains based on Tendermint consensus, a variety of attacks against the network are possible if an attacker acquire 1/3+ or 2/3+ of all bonded tokens. Since the market cap of newly created chains could be relatively low, an attacker could realistically acquire sufficient tokens to pass these thresholds. As a solution, CCV allows the newly created chains to use validators that have stake on chains with a much larger market cap and, as a result, increase the cost an attacker would have to pay. 

> Moreover, CCV enables *hub minimalism*. In a nutshell, hub minimalism entail keeping a hub in the Cosmos network (e.g., the Cosmos Hub) as simple as possible, with as few features as possible in order to decrease the attack surface. CCV enables moving distinct features (e.g., DEX) to independent chains that are validated by the same set of validators as the hub. 

Reviewer 2: Typo: " ..., hub minimalism entails ...".

> **Versioning**: Note that CCV will be developed progressively. 
> - The V1 release will require the validator set of a consumer chain to be entirely provided by the provider chain. In other words, once a provider chain agrees to provide security to a consumer chain, the entire validator set of the provide chain MUST validate also on the consumer chain. As a result, the consumer chain can be thought of as having the same security guarantees as the provider chain. 
> - The V2 release will allow the provider chain validators to opt-in to participate as validators on the consumer chain. It is up to each consumer chain to establish the benefits that provider chain validators receive for their participation.
> 
> For more details on the planned releases, take a look at the [Interchain Security light paper](https://github.com/cosmos/gaia/blob/main/docs/interchain-security.md#the-interchain-security-stack).

### Definitions and Overview

This section defines the new terms and concepts introduced by CCV and provides an overview of CCV.

**Provider Chain**: The blockchain that provides security, i.e., manages the validator set of the consumer chain.

> **Consumer Chain**: The blockchain that consumes security, i.e., enables the provider chain to correctly manage its validator set.

Reviewer 2: Maybe we could remove "correctly" from the sentence.
It is defined what that means.

> Note that in the current version the validator set of the consumer chain is entirely provided by the provider chain.

**CCV Module**: The module that implements the CCV protocol.

> **CCV Channel**: A unique, ordered IBC channel (as define in [ICS 4](../../core/ics-004-channels-and-packets-semantics)) that is used by the two CCV modules to exchange messages.

Reviewer 2: Typo: "... (as defined ...").

> Note that the IBC handler interface & IBC relayer module interface are as defined in [ICS 25](../../core/ics-025-handler-interface) and [ICS 26](../../core/ics-026-routing-module), respectively.

CCV must handle the following types of operations:

> - **Channel Initialization**: Create a unique, order IBC channel between the provider chain and the consumer chain.

Reviewer 2: Typo: "... a unique, ordered ...".

  - The channel is not built on top of the client created for this consumer chain.
  - Another CCV channel for this consumer chain already exists.

```typescript
// implements the ICS26 interface
function onChanOpenAck(
  portIdentifier: Identifier,
  channelIdentifier: Identifier,
  counterpartyVersion: string) {
    // the channel handshake MUST be initiated by consumer chain
    abortTransactionUnless(FALSE)
}
```
- **Initiator:**
  - The IBC module on the provider chain.
- **Expected precondition:** 
  - The IBC module on the provider chain received a `ChanOpenAck` message on a port the provider chain CCV module is bounded to.
- **Expected postcondition:** 
  - The state is not changed.
- **Error condition:**
  - Invoked on the provider chain.

```typescript
// implements the ICS26 interface
function onChanOpenConfirm(
  portIdentifier: Identifier,
  channelIdentifier: Identifier) {
    // get the client state associated with this client ID in order 
    // to get access to the consumer chain ID
    clientId = getClient(channelIdentifier)
    clientState = clientKeeper.GetClientState(clientId)

    // Verify that there isn't already a CCV channel for the consumer chain

> 	  // If there is, then close the channel.

Reviewer 2: We can fix the indentation for this comment.

  - `providerChannel` is not set or `providerChannel` matches the ID of the channel the `ChanCloseInit` message was received on.

```typescript
function onChanCloseConfirm(
  portIdentifier: Identifier,
  channelIdentifier: Identifier) {
    // do nothing
}
```
- **Initiator:**
  - The IBC module on the consumer chain.
- **Expected precondition:** 
  - The IBC module on the consumer chain received a `ChanCloseConfirm` message on a port the consumer chain CCV module is bounded to.
- **Expected postcondition:** 
  - The state is not changed.
- **Error condition:**
  - None.

#### Packet Relay

**On the provider chain**

```typescript
// implements the ICS26 interface
function onRecvPacket(packet: Packet): Packet {
  switch typeof(packet.data) {
    // the provider chain receives no packets
    default:
      // unknown packet type
      return NewErrorAcknowledgement()
  }    
}
```
- **Initiator:**
  - The IBC module on the provider chain.
- **Expected precondition:** 
  - The IBC module on the provider chain received a packet on a channel owned by the provider chain CCV module.
- **Expected postcondition:** 
  - The state is not changed.
- **Error condition:**

>   - The packet type is unknown.

Reviewer 2: I think that a reception of any packet (i.e., any type of a packet) is the error condition.
In other words, the reception of VSCPacketData on the provider chain should be an error, even though VSCPacketData is "known".
This is just a minor clarification, in my opinion.

Author: I replaced with unexpected.

```typescript
// implements the ICS26 interface
function onRecvPacket(packet: Packet): Packet {
  switch typeof(packet.data) {
    case VSCPacketData:
      return onRecvVSCPacket(packet)
    default:
      // unknown packet type
      return NewErrorAcknowledgement()
  }
}
```
- **Initiator:**
  - The IBC module on the consumer chain.
- **Expected precondition:** 
  - The IBC module on the consumer chain received a packet on a channel owned by the consumer chain CCV module.
- **Expected postcondition:** 
  - If the packet is a `VSCPacket`, the `onRecvVSCPacket` method is invoked.
- **Error condition:**
  - The packet type is unknown.

```typescript
// implements the ICS26 interface
function onAcknowledgePacket(packet: Packet) {
  switch typeof(packet.data) {
    // the consumer chain sends no packets
    default:
      // unknown packet type
      abortTransactionUnless(FALSE)
  }
}
```
- **Initiator:**
  - The IBC module on the consumer chain.
- **Expected precondition:** 
  - The IBC module on the consumer chain received an acknowledgement on a channel owned by the consumer chain CCV module.
- **Expected postcondition:** 
  - The state is not changed.
- **Error condition:**

>   - The acknowledgement is for an unknown packet.

Reviewer 2: Same comment as the one right above.

```typescript
// implements the ICS26 interface
function onTimeoutPacket(packet Packet) {
  switch typeof(packet.data) {
    // the consumer chain sends no packets
    default:
      // unknown packet type
      abortTransactionUnless(FALSE) 
  }
}
```
- **Initiator:**
  - The IBC module on the consumer chain.
- **Expected precondition:** 
  - The IBC module on the consumer chain received a timeout on a channel owned by the consumer chain CCV module.
  - The Correct Relayer assumption is violated.
- **Expected postcondition:** 
  - The state is not changed.
- **Error condition:**

>   - The timeout is for an unknown packet.

Reviewer 2: Same comment as the one right above.

### Definitions and Overview
[&uparrow; Back to Outline](#outline)

This section defines the new terms and concepts introduced by CCV and provides an overview of CCV.

**Provider Chain**: The blockchain that provides security, i.e., manages the validator set of the consumer chain.

**Consumer Chain**: The blockchain that consumes security, i.e., enables the provider chain to manage its validator set.

> Note that in the current version the validator set of the consumer chain is entirely provided by the provider chain.

**CCV Module**: The module that implements the CCV protocol.

**CCV Channel**: A unique, ordered IBC channel (as defined in [ICS 4](../../core/ics-004-channels-and-packets-semantics)) that is used by the two CCV modules to exchange messages.

> Note that the IBC handler interface & IBC relayer module interface are as defined in [ICS 25](../../core/ics-025-handler-interface) and [ICS 26](../../core/ics-026-routing-module), respectively.

CCV must handle the following types of operations:
- **Channel Initialization**: Create a unique, ordered IBC channel between the provider chain and the consumer chain.
- **Validator Set Update**: It is a two-part operation, i.e., 

>   - update the validator set of the consumer chain based on the information received from the Staking module on the amount of tokens bonded by validators on the provider chain;

Author: I've added the Validator Set Update figure after the definition of VSC. The Channel Initialization figure is still before though. In the description of the figure we do refer to the next section when we mention VSC, but I agree that it's not ideal. 

I find that by putting the VSC definition in the enumeration of operation types, would take away from its significance.

Note that the "*FIRST*" keyword in the above description ensures the uniqueness of the IBC channel.

<!-- omit in toc -->
#### Validator Set Update Overview

The following Figure shows an overview of the Validator Set Update operation of CCV. 

![Validator Set Update Overview](./ccv-vsc-overview.png?raw=true)

**Validator Set Change (VSC)**: A change in the validator set of the provider chain that must be reflected in the validator set of the consumer chain. It consists of a batch of validator updates, i.e., changes in the voting power granted to validators on the provider chain and, due to CCV, also on the consumer chain.

> **Background**: In the context of single-chain validation, the changes of the validator set are triggered by the Staking module. For more details, take a look at the [Cosmos SDK documentation](https://docs.cosmos.network/master/modules/staking/). 

**Matured VSC**: A VSC that has matured on the consumer chain, i.e., a certain period of time, known as the *unbonding period* (i.e., `UnbondingPeriod`) has elapsed since the VSC was received by the consumer chain. 

> > TODO: The unbonding period for a VSC should be started from the time of applying the VSC. At the moment it is from the time of receiving the VSC.

Author: Also for being accurate. At the moment it doesn't really matter because the difference between receiving and applying a VSC is maximum the time to create a new block on the consumer chain, which is a few seconds. However, if this is to be change in the future (e.g., introduction of epochs), then it might be relevant.

> > TODO: The unbonding period for a VSC should be started from the time of applying the VSC. At the moment it is from the time of receiving the VSC.

Author: `unbondingTime := ctx.BlockTime().Add(types.UnbondingTime)` ->  it doesn't matter when this is called within a block

> **Intuition**: Every VSC consists of a batch of validator updates, some of which can be decreases in the voting power granted to validators. These decreases are a consequence of unbonding operations on the provider chain, which MUST NOT complete before reaching maturity on both the provider and all the consumer chains. Thus, a VSC reaching maturity on a consumer chain means that all the unbonding operations that resulted in validator updates included in that VSC have matured on the consumer chain.

> > **Background**: An *unbonding operation* is any operation of unbonding an amount of the tokens a validator bonded and, as a result, was granted voting power. Unbonding operations have two components: 

Author: I don't find it clearer since "**its** voting power" refers to the validator mentioned afterwards. 

What about:
> **Background**: An *unbonding operation* is any operation of unbonding an amount of the tokens a validator bonded. Note that the bonded tokens correspond to the validator's voting power. Unbonding operations have two components:

> - The *initiation*, e.g., a delegator request their delegated tokens to be unbonded. The initiation of an operation of unbonding an amount of the tokens a validator bonded results in a change in the voting power of that validator.
> - The *completion*, e.g., the tokens are actually unbonded and transferred back to the delegator. To complete, unbonding operations must reach *maturity*, i.e., `UnbondingPeriod` must elapse since the operations were initiated.

In the context of VSCs, the CCV module enables the following functionalities:
  - On the provider chain, 

>     - **request** VSCs to the consumer chain, to update its validators set, according to the validator set of the provider chain;

Author: I agree. Request to is not correct English. Request from would be correct, but *the provider chain requesting VSCs from the consumer chain* implies that the consumer chain will send back VSCs, which makes the entire thing very confusing. 

The actual problem here is the *request* word. I suggest changing it for one of the following:
- *provide* VSCs to the consumer chain
- *issue* VSCs  to the consumer chain

I would go for provide since this is what a *provider* does :)

>     - **request** VSCs to the consumer chain, to update its validators set, according to the validator set of the provider chain;

Reviewer 3: I am shocked but Oxford's dictionary actually allows "provide something to someone". So let's do that ;-)

The following assumption defines the guarantee on unbonding operations that CCV expects to be provided by the Staking module. 

> - **Unbonding Timeliness** Let `UO` be an unbonding operation that results in a validator update included in a VSC `C` requested by the provider chain. Then, `UO` MUST NOT complete on the provider chain before the provider chain registers notifications of `C`'s maturity from all consumer chains. 

Author: The definition of `UO(C)` is not accurate: An unbonding operation does not start with requesting a VSC. 

An unbonding operation leads to a validator update that will be included eventually in a VSC requested by the provider chain.

> The following assumption defines the guarantee on unbonding operations that CCV expects to be provided by the Staking module. 
> - **Unbonding Timeliness** Let `UO` be an unbonding operation that results in a validator update included in a VSC `C` requested by the provider chain. Then, `UO` MUST NOT complete on the provider chain before the provider chain registers notifications of `C`'s maturity from all consumer chains. 

Reviewer 3: I see. What about

```suggestion
We define an "unbonding operation" `UO` that starts with an unboding transaction being executed and ends with the event that returns the corresponding stake. Let `C(UO)` be the VSC that contains the validator update corresponding to `UO`.

- **Staking Safety** For every `UO` it holds that `UO` MUST NOT end on the provider chain before the provider chain registers notifications of `C(UO)`'s maturity from all consumer chains. 
```

### Desired Properties
[&uparrow; Back to Outline](#outline)

The following properties are concerned with **one provider chain** providing security to **multiple consumer chains**. Between the provider chain and each consumer chain, a separate CCV channel is established. 

First, we define the properties for the CCV channels. Then, we define the guarantees provided by CCV.

- **Channel Uniqueness**: The channel between the provider chain and a consumer chain MUST be unique.
- **Channel Validity**: If a packet `P` is received by one end of the channel, then `P` MUST have been sent by the other end of the channel.
- **Channel Order**: If a packet `P1` is sent over the channel before a packet `P2`, then `P2` MUST NOT be received by the other end of the channel before `P1`. 
- **Channel Liveness**: Every packet sent over the channel MUST eventually be received by the other end of the channel. 

> CCV provides the following system invariants:

Author: The entire spec is targeting only V1. The other versions (i.e., V2) are only mentioned in the motivation. It's hard to decide now what's targeting only V1 and what will also work for V2, since we don't have a clear design for V2.

> CCV provides the following system invariants:

Reviewer 3: OK. Then let's keep it as it is and see in V2 what needs to be specialized/generalized/

As a consequence, CCV can rely on the following property:
- **Validator Update Inclusion**: Let `U1` and `U2` be two validator updates targeting the same validator `V`. If `U1` occurs before `U2`, then `U2` sums up all the changes of the voting power of `V` that are summed up by `U1`, i.e., 
  - `U1 = c_1+c_2+...+c_i` and
  - `U2 = c_1+c_2+...+c_i+c_(i+1)+...+c_j`.
 
The Validator Update Inclusion property enables CCV to aggregate multiple VSCs. It is sufficient for the consumer chains to apply only the last update per validator. Since the last update of a validator *includes* all the previous updates of that validator, once it is applied, all the previous updates are also applied.

---

<!-- omit in toc -->

> #### Staking Module Interface

Author: Same as above.

- `unbondingPackets: List<(Packet, Time)>` is a list of `(packet, unbondingTime)` tuples, where `packet` is a received `VSCPacket` and `unbondingTime` is the packet's unbonding time. The list is used to keep track of when unbonding operations are matured on the consumer chain. It exposes the following interface:
  ```typescript
  interface List<(Packet, Time)> {
    // add a packet with its unbonding time to the list
    Add(packet: Packet, unbondingTime: Time)

    // return the list sorted by the unbonding time
    SortedByUnbondingTime(): List<(Packet, Time)>

    // remove (packet, unbondingTime) from the list
    Remove(packet: Packet, unbondingTime: Time)
  }
  ```
 
### Sub-protocols

> TODO: What about `EmitEvent()`? There are some events emitted throughout the code, but none of them are specified. 

To express the error conditions, the following specification of the sub-protocols uses the exception system of the host state machine, which is exposed through two functions (as defined in [ICS 24](../../core/ics-024-host-requirements)): `abortTransactionUnless` and `abortSystemUnless`.

#### Initialization
[&uparrow; Back to Outline](#outline)

> The *initialization* sub-protocol enables a provider chain and a consumer chain to create a CCV channel -- a unique, ordered IBC channel for exchanging messages. As a prerequisite, the initialization sub-protocol MUST create two IBC clients, one on the provider chain to the consumer chain and one on the consumer chain to the provider chain. This is necessary to verify the identity of the two chains (as long as the clients are trusted). 

Reviewer 3: ```suggestion
The *initialization* sub-protocol enables a provider chain and a consumer chain to create a CCV channel -- a unique, ordered IBC channel for exchanging packets. As a prerequisite, the initialization sub-protocol MUST create two IBC clients, one on the provider chain to the consumer chain and one on the consumer chain to the provider chain. This is necessary to verify the identity of the two chains (as long as the clients are trusted). 
```
I guess it should be "packets". If not, please ignore...

**On the provider chain**

> 

Reviewer 3: I would suggest to add tags for each of the following functions for two reasons
1. typographically they highlight that something important is happening at this point (the start of the individual functions is less visible than the bold bullet points below)
2. for asynchronous discussion having links to point to was super useful in the past. (With the #### section we can put direct links to this spot)
 
```suggestion

#### **[CCV-PCF-INITG.1]**
```
- PCF for provider chain function, CCF for consumer chain function

```typescript
// implements the AppModule interface
function InitGenesis(state: ProviderGenesisState) {
  // bind to ProviderPortId port 
  err = portKeeper.BindPort(ProviderPortId)
  // check whether the capability for the port can be claimed
  abortSystemUnless(err == nil)

  foreach cs in state.clientStates {
    chainToChannel[cs.chainId] = cs.channelId
    channelToChain[cs.channelId] = cc.chainId
    channelStatus[cs.channelId] = cc.status
  }
}
```
- **Initiator:**
  - ABCI when the provider chain is started.
- **Expected precondition:** 

>   - The provider chain is started for the first time.

Reviewer 3: Does it fail if this is not the case?

Author: I don't think so. The error condition is below. 

From https://docs.cosmos.network/v0.44/building-modules/genesis.html:
> The InitGenesis method is executed during InitChain when the application is first started.

From https://docs.cosmos.network/v0.44/core/baseapp.html#initchain:
> The InitChain ABCI message (opens new window) is sent from the underlying Tendermint engine when the chain is first started.

Reviewer 3: So it is basically an assumption about CosmosSDK?

Author: More of a fact of how Cosmos SDK works as long as the blockchain is valid. 

I've changed it to this:
- Initiator:
  - ABCI.
- Expected precondition: 
  - An `InitChain` message is received from the consensus engine; the `InitChain` message is sent when the provider chain is first started.

- **Expected postcondition:**
  - The capability for the port `ProviderPortId` is claimed.
  - For each consumer state in the `ProviderGenesisState`, the initial state is set, i.e., the following mappings `chainToChannel`, `channelToChain`, `channelStatus` are set.
- **Error condition:**
  - The capability for the port `ProviderPortId` cannot be claimed.

```typescript
// implements governance proposal Handler 
function CreateConsumerChainProposal(p: CreateConsumerChainProposal) {
  if BlockTime() > p.spawnTime {

>     // create consumer chain

Reviewer 3: What is the meaning of the comment without any code below? Is something supposed to happen somewhere here?

Author: Nope, just a bad comment :(

kind: 
author: 
created: 
modified: 
---

## Synopsis

This standard document specifies packet data structure, state machine handling logic, and encoding details for Cross-Chain Validation (CCV). Note that CCV is the specific IBC level protocol that enables *Interchain Security*, a Cosmos-specific category of *Shared Security*.

At a high level, CCV enables a *provider chain* (e.g., the Cosmos Hub) to provide *security* to a *consumer chain*. Namely, the validators of the consumer chain are chosen based on the staking tokens that have been delegated to them on the provider chain. When these validators misbehave on the consumer chain, their stake on the provider chain is slashed. As a result, the security gained from the value of the stake bonded on the provider chain is shared with the consumer chain. For more details, take a look at the [Interchain Security light paper](https://github.com/cosmos/gaia/blob/main/docs/interchain-security.md).

The exchange of messages between the provider and consumer chains is done through the IBC protocol over an *unique*, *ordered* channel. 

> Throughout this document, we will use the terms chain and blockchain interchangeably.

### Motivation

> CCV is a primitive (i.e., a building block) that enables arbitrary shared security models: The security of a chain can be composed of security transferred from multiple provider chains including the chain itself (a consumer chain can be its own provider). As a result, CCV enables chains to borrow security from more established chains (e.g., Cosmos Hub), in order to boost their own security, i.e., increase the cost of attacking their networks. 

Author: I took the motivation from the [Interchain Security light paper](https://github.com/cosmos/gaia/blob/main/docs/interchain-security.md) and from Ethan's coindesk talk https://drive.google.com/file/d/1WSqtT20-Osgekzw4EhPhfdbglsfn6oEq/view?ts=618ab27c. 
- By arbitrary I mean that since CCV is a building block, it enables users to build many flavors of security models.
- The "The security of a chain can be composed of security transferred from multiple provider chains including the chain itself (a consumer chain can be its own provider)." statement refers to future versions of CCV. Basically I want to state the vision. 
 
Although this spec is targeting V0, I wanted the motivation to be about CCV in general. Should we make it just about V0, i.e., what's the motivation behind the consumer chain getting its entire set from a single provider chain?

- **Undelegations**: Delegators request their delegated tokens to be *unbonded* (i.e., no longer bonded). Every undelegation (or unbonding delegation) has two effects: first, the unbonded tokens are no longer contributing to the validator's voting power; and second, after `UnbondingPeriod` has elapsed from the current time, the unbonding *matures* and the unbonded tokens are transferred back to the delegator.
- **Redelegations**: Delegators request their delegated tokens to be be instantly redelegated from a source validator to a different validator (the destination validator). Every redelegation has two effects: first, the redelegated tokens are no longer contributing to the source validator's voting power and are contributing to the destination validator's voting power; and second, after `UnbondingPeriod` has elapsed from the current time, the redelegation *matures* and the redelegated tokens can no longer be slashed for misbehaviors of the source validator (note that these misbehaviors must have happened before the redelegation was initiated).

> 

Author: I took everything concerning slashing and distribution out of the current draft.

> 

Author: TODO: Add slashing as a trigger for validator updates

In addition, every validator can be in one of the following three states: 
- *Unbonded*: Validators not in the validator set.
- *Bonded*: Once a validator receives sufficient bonded tokens they automatically join the validator set.

> - *Unbonding*: When a validator leaves the validator set, either by choice or not, an unbonding of all their bonded tokens begins. After `UnbondingPeriod` has elapsed, the unbonding *matures* and the unbonded tokens can be transferred back to their owners.

Author: What about the following:
*Unbonding*: A validator that left the validator set at a time `T` with the condition that less than `UnbondingPeriod` has elapsed since `T`.

> - *Unbonding*: When a validator leaves the validator set, either by choice or not, an unbonding of all their bonded tokens begins. After `UnbondingPeriod` has elapsed, the unbonding *matures* and the unbonded tokens can be transferred back to their owners.

Reviewer: ok.

Note that delegations, undelegations, and redelegations targeting unbonded validators have no effect on the validator set. 

### Definitions

The following Figure shows an overview of CCV together with the new terms and concepts introduced.

![CCV Overview](./ccv-vsc-overview.png?raw=true)

**Provider Chain**: The blockchain that provides security, i.e., manages the validator set of the consumer chain.

**Consumer Chain**: The blockchain that consumes security, i.e., enables the provider chain to correctly manage its validator set.

> Note that in the current version the validator set of the consumer chain is entirely provided by the provider chain. 

**Validator Set Change (VSC)**: A change in the validator set of the provider chain that must be reflected in the validator set of the consumer chain. It consists of a batch of validator updates, i.e., changes in the voting powers granted to validators on the provider chain and, due to CCV, also on the consumer chain.

> **CCV Unbonding Operation**: An operation of unbonding an amount of the tokens a provider chain validator bonded and, as a result, was granted voting powers on the consumer chain. 

Author: It makes it easier to define the desired properties, by distinguishing between the effects of unbonding operations already existent in single-chain staking (e.g., adding entries in a time based UnbondingDelegation queue) and the effects on the consumer chain (i.e., sending back an ACK after `UnbondingPeriod`). 
- **CCV unbonding operations** refer to the operations of unlocking the stake bonded in order to get voting powers on the consumer chain.
- **Unbonding operations** refer to the operations of unlocking the stake bonded in order to get voting powers on the provider chain.

> Note that in the current version there is no difference between CCV unbonding operations and single-chain unbonding operations.

> **Matured CCV Unbonding Operation**: A CCV unbonding operation that has matured on the consumer chain. 

Author: It makes it easier to define the desired properties by not having to say all the time that a CCV unbonding operation has matured on the consumer chain. We could add it as a note, i.e., 
> Note that by saying that a CCV unbonding operation has matured, we mean that it has matured on the consumer chain.

**CCV Module**: The module that implements the CCV protocol. It enables the following functionalities:
  - On the provider chain, 
    - **request** VSCs in order to keep an up-to-date view of which provider chain validators can produce blocks on the consumer chain;
    - **register** maturity notifications in order to ensure that the unbonding operations for validators that produce blocks on the consumer chain do not complete prematurely.
  - On the consumer chain,
    - **apply** the VSCs requested by the provider chain to the validator set of the consumer chain; 

>     - **notify** the provider chain when CCV unbonding operations have matured.

Author: But I do believe that CCV needs to modify the existing unbonding operations, i.e., the staking hooks and postponing completion on the provider chain until receiving ACKs from the consumer chain. We could clearly redefine **notify** as you suggested, but it's more verbose. This doesn't have a significant impact here, but it would make the properties much more difficult to follow.

> Note that for brevity, we say that "the provider chain requests a CCV unbonding operation" instead of "the provider chain requests a VSC that contains the update of a validator's voting power caused by a CCV unbonding operation". 

**CCV Channel**: A unique, ordered IBC channel (as define in [ICS 4](../../core/ics-004-channels-and-packets-semantics)) that is used by the two CCV modules to exchange messages.

The IBC handler interface & IBC relayer module interface are as defined in [ICS 25](../../core/ics-025-handler-interface) and [ICS 26](../../core/ics-026-routing-module), respectively.

### Assumptions

This standard relies on the following system assumptions:

- **Valid Blockchain**: Both the provider and the consumer chains are *valid*. This means that the protocols are executed correctly and the underlying consensus algorithm satisfies both safety and liveness properties. For more details, take a look at the [Tendermint Paper](https://arxiv.org/pdf/1807.04938.pdf).

> - **Correct Relayer**: There is at least one *correct* relayer between the provider and consumer chains -- every packet sent on the CCV channel is relayed to the receiving end before the packet timeout elapses. Clearly, the CCV protocol is responsible of setting the packet timeouts (i.e., `timeoutHeight` and `timeoutTimestamp`) such that the Correct Relayer assumption is feasible. 

Author: The packet timeouts do not refer to a time period, but rather to **a time instance** on the destination chain. These are the defs of the packet timeouts (cf. [ICS 4](https://github.com/cosmos/ibc/blob/master/spec/core/ics-004-channel-and-packet-semantics/README.md#definitions)):
- The `timeoutHeight` indicates **a consensus height on the destination chain after which** the packet will no longer be processed, and will instead count as having timed-out.
- The `timeoutTimestamp` indicates **a timestamp on the destination chain after which** the packet will no longer be processed, and will instead count as having timed-out.

As a result, `tH` and `tO` will not help. The application on the sender chain would need to know the current block `B` and time `T` on the destination chain in order to set packets timeouts to values accordingly, i.e., `timeoutHeight = B + tH` and `timeoutTimestamp = T + tO`. 

In practice, we just need to set `timeoutTimestamp` to a value far enough in the future to make sure packets cannot timeout. This is actually how the current implementation works, i.e., `timeoutTimestamp = BlockTime on sender chain + 4 weeks` (see https://github.com/cosmos/interchain-security/blob/7865e670550d2894b6d4ffa4f4bf8d999d2fc923/x/ccv/parent/types/utils.go#L8)

 
> **Discussion**: IBC relies on timeouts to signal that a sent packet is not going to be received on the other end. Once an ordered IBC channel timeouts, the channel is closed (see [ICS 4](../../core/ics-004-channels-and-packets-semantics)). The **Correct Relayer** assumption is necessary to ensure that the CCV channel **cannot** ever timeout and, as a result, cannot transit to the closed state. 
> 
> In practice, the **Correct Relayer** assumption is realistic since any validator could play the role of the relayer and it is in the best interest of correct validators to successfully relay packets.

> TODO: Can we provide an estimate of the expected block rate based on the timeout assumption of the relayer?

Although the following assumptions follow directly from the Valid Blockchain assumption, it is important to clearly state what is CCV expecting from the other modules in the system (e.g., the Staking module).

- **ValidatorUpdate-to-VSC Validity**: The validator updates in any VSCs requested by the provider chain MUST reflect the updates in the voting powers of validators on the provider chain.

> - **ValidatorUpdate-to-VSC Order**: Let U1 and U2 be two updates in the voting powers of validators on the provider chain. Then, U1 and U2 MUST NOT be included *out of order* in VSCs requested by the provider chain. 

Author: These assumptions refer to the connection between updates in the voting power of validators on the provider chain (denoted by U) and VSCs (validator set changes) requested by the provider chain. The updates happen in the staking module, while request VSCs is a CCV functionality.

- **ValidatorUpdate-to-VSC Liveness**: Every update in the voting powers of validators on the provider chain MUST eventually be included in a VSC requested by the provider chain. 

### Desired Properties

First, we define the properties for the CCV channel. Then, we define the guarantees provided by CCV.

- **Channel Uniqueness**: The provider and consumer chains communicate over a unique channel.
- **Channel Validity**: If a packet P is received by one end of the channel, then P MUST have been sent by the other end of the channel.
- **Channel Order**: If a packet P1 is sent over the channel before a packet P2, then P2 MUST NOT be received by the other end of the channel before P1. 
- **Channel Liveness**: Every packet sent over the channel MUST eventually be received by the other end of the channel. 

> Note that the following properties are concerned with only one provider chain and one consumer chain. 

> > TODO: Do we need properties for multiple consumer chains? What the interaction between two consumer chains connected to the same provider chain?

Author: Yes, there is a unique channel. However, are there other CCV properties regarding the provider chain managing multiple consumer chains? At the moment, all the properties are from the perspective of only one provider chain and one consumer chain. That was the reason for adding the TODO. If there are no other properties, we can just remove it.

> > TODO: Do we need properties for multiple consumer chains? What the interaction between two consumer chains connected to the same provider chain?

Reviewer: There will be a slight change, e.g., we should call VSC to be mature if the unbonding time of each consumer chain elapsed.

CCV is based on the following system invariants:

> - The voting power granted to validators on the consumer chain corresponds to the amount of tokens bonded by these validators on the provider chain.

Author: Is that true? How is the voting power of a validator computed at the moment? Is it based only on the amount of tokens bonded by that validator or is it computed relatively to the total amount of tokens bonded by the validator set? If it's the former, than the invariant  "The voting power of any validator on a consumer chain, equals its voting power on the producer chain." holds. If it's the latter, then it will not hold if the consumer chain can remove validators from the validator set when jailing them (as we discussed in the meeting).

> - The voting power granted to validators on the consumer chain corresponds to the amount of tokens bonded by these validators on the provider chain.

Author: Corresponds is vague ...

- The unbonding of any amount of the tokens bonded on the provider chain by validators validating on the consumer chain must mature on the consumer chain.

For these invariants to hold, CCV makes some assumptions about the other modules in the system (as defined in the [Assumptions](#assumptions) Section). 

In addition, CCV must provide the following guarantees to the rest of the system.

The following properties define the guarantees of *applying* on the provider chain VSCs *requested* by the provider chain.
- **VSC Validity**: Every VSC applied by the consumer chain MUST be requested by the provider chain.

> - **VSC Order**: If a VSC C1 containing an update of a validator V1 is requested by the provider chain before a VSC C2 containing an update of a validator V2, but not an update of V1, then the consumer chain MUST NOT apply the update of V2 before the update of V1. (*Note:* The order of updates within a VSC does not matter.)

Author: In my view, this is the flow of validator updates during CCV: 
Provider_Staking (initiate updates) -> Provider_Registry (batch updates into VSC) -> Provider_CCV (request VSC) -> CCV Channel (relay VSCs) -> Consumer_CCV (aggregate VSCs and apply at EndBlock, and send ACK after unbonding period) -> CCV Channel (relay ACKs) -> Provider_CCV (receives ACKs and notifies Provider_Staking of maturity via hooks). 

The current properties are addressing the Provider_CCV (request VSC) -> ... -> Provider_CCV (receives ACKs and notifies Provider_Staking)  part of the sequence. 

The **ValidatorUpdate-to-VSC** assumptions are addressing the Provider_Staking (initiate updates) -> Provider_Registry (batch updates into VSC) -> Provider_CCV (request VSC)  part of the sequence. 

Note that none of the  **ValidatorUpdate-to-VSC** are used for proving correctness since the properties are from the perspective of the CCV module. I added them for completion.

> Marius: If I were to define VSC Order just for *disjoint* VSCs, it will not address the order between two sets that are not disjoint. 
- **VSC Liveness**: If the provider chain requests a VSC containing an update of a validator V and it does not request any other VSCs containing updates of V, then the consumer chain MUST eventually apply the update of V.

The following properties define the guarantees of *registering* on the provider chain maturity notifications of CCV unbonding operations *requested* by the provider chain.
- **CCV Unbonding Validity**: If the provider chain registers a maturity notification of a CCV unbonding operation U, then the provider chain MUST have requested it. 

> - **CCV Unbonding Timeliness**: If the provider chain requests a CCV unbonding operations U, then the provider chain MUST NOT register a maturity notification of U before `UnbondingPeriod` has elapsed since requesting it.

Author: There is the following note in the Definitions Section:
> Note that for brevity, we say that "the provider chain requests a CCV unbonding operation" instead of "the provider chain requests a VSC that contains the update of a validator's voting power caused by a CCV unbonding operation". 
The reason is to make the properties easier to read, but if you say it's confusing, maybe it's not achieving its intended purpose. Should I remove it?
Actually, the val set upd is a consequence of an unbonding operation. Unbonding operations have two components: the initiation, e.g., when a delegator requests to unlock its tokens, and the completion, when the tokens are actually unlocked. The initiation leads to a change in the voting powers, which leads to the CCV module on the provider chain requesting a VSC; after `UnbondingPeriod` elapses on the consumer chain, it sends an ACK, which enables the provider chain to complete the unboding. Note that the provider chain still waits for the unbonding to mature locally.

#### CCV Packets

The structure of the packets sent through the CCV channel is defined by the `Packet` interface in [ICS 4](../../core/ics-004-channels-and-packets-semantics). Packets are acknowledged by the remote side by sending back an `Acknowledgement` that contains either a result, created with `NewResultAcknowledgement()`, or an error, created with `NewErrorAcknowledgement()`. 

The following packet data types are required by the CCV module:
- `VSCPacketData` contains a list of validator updates, i.e., 
    ```typescript
    interface VSCPacketData {
      updates: [ValidatorUpdate]
    }
    ```
> Note that for brevity we use e.g., `VSCPacket` to refer to a packet with `VSCPacketData` as its data.

#### CCV State

This section describes the internal state and data structures of the CCV module

**Common state across both provider and consumer chains**
- `channelStatus: Map<Identifier, ChannelStatus>` is a mapping from CCV channel IDs to `ChannelStatus`s, used to indicate in which state a CCV channel is. `ChannelStatus` is defined as
  ```typescript

>   enum ChannelStatus {

Author: I would suggest to use tags instead in order to enable Shon's tool. However, if we are not sure about tagging yet, we could just add links for now. What do you think?

>   enum ChannelStatus {

Reviewer: let's start with links, and add tags next.

>   enum ChannelStatus {

Author: Code moves too fast to keep links up to date. We'll add tags.

- **Error condition:**
  - None.

```typescript
function UnbondMaturePackets(): Error {
  if providerChannel == "" {
    // provider channel not set
    return Error()
  }
  // get the current time
  currentTime = BlockTime()

  foreach (packet, unbondingTime) in unbondingPackets.SortedByUnbondingTime() {
    if currentTime >= unbondingTime {
      // send acknowledgement to the provider chain
      channelKeeper.WriteAcknowledgement(providerChannel, packet, NewResultAcknowledgement())
            
      // remove entry from the list
      unbondingPackets.Remove(packet, unbondingTime)
    } else {
      // stop loop
      break
    }
  }
  return nil
}
```
- **Initiator:** `EndBlock()` method.
- **Expected precondition:**
  - `providerChannel` is set.
- **Expected postcondition:**
  - For each `(packet, unbondingTime)` in the sorted (by unbonding times) list of unboding packet
    - if `BlockTime() >= unbondingTime`, the packet is acknowledged and the tuple is removed from `unbondingPackets`;
    - otherwise, stop the loop.
- **Error condition:**
  - If the precondition is violated.

### Properties & Invariants

#### Correctness Reasoning

> - **Channel Uniqueness:** The provider chain sets the CCV channel when receiving (from the consumer chain) the first `ChanOpenConfirm` message and it marks the channel as `INVALID` when receiving any subsequent `ChanOpenConfirm` messages (cf. Valid Blockchain). Similarly, the consumer chain sets the CCV channel when receiving the first `VSCPacket` and ignores any packets received on different channels (cf. Valid Blockchain). 

Author: I'm not sure I understand what you mean. Maybe we can move this discussion on Slack.

- **Channel Validity:** Follows directly from the Valid Blockchain assumption.

- **Channel Order:** The provider chain accepts only ordered channels when receiving a `ChanOpenTry` message (cf. Valid Blockchain). Similarly, the consumer chain accepts only ordered channels when receiving `ChanOpenInit` messages (cf. Valid Blockchain). Thus, the property follows directly from the fact that the CCV channel is ordered. 

> - **Channel Liveness:** The property follows from the Correct Relayer assumption. 

Author: I added a discussion after the definition of the Correct Relayer assumption on how should CCV choose the packet timeout so the assumption holds. I find it fits better there, since it's a discussion of an assumption. Here we just reason about the correctness of the Channel Liveness property.
